src/main/java/com/stripe/net/StripeRequest.java

```
@@ -1,32 +1,168 @@
+      this.url = buildURL(method, url, params);
+      this.content = buildContent(method, params);
+      this.headers = buildHeaders(method, this.options);
+    } catch (IOException e) {
+      throw new ApiConnectionException(
+          String.format(
+              "IOException during API request to Stripe (%s): %s "
+                  + "Please check your internet connection and try again. If this problem persists,"
+                  + "you should check Stripe's service status at https://twitter.com/stripestatus,"
+                  + " or let us know at support@stripe.com.",
+              Stripe.getApiBase(), e.getMessage()),
+          e);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private static URL buildURL(
+      ApiResource.RequestMethod method, String spec, Map<String, Object> params)
+      throws IOException {
+    StringBuilder sb = new StringBuilder();
+
+    sb.append(spec);
+
+    if ((method != ApiResource.RequestMethod.POST) && (params != null)) {
+      String queryString = FormEncoder.createQueryString(params);
+      if (queryString != null && !queryString.isEmpty()) {
+        sb.append("?");
+        sb.append(queryString);
+      }
+    }
+
+    URL url;
+    String customUrlStreamHandlerClassName =
+        System.getProperty(CUSTOM_URL_STREAM_HANDLER_PROPERTY_NAME, null);
+    if (customUrlStreamHandlerClassName != null) {
+      // instantiate the custom handler provided
+      Class<URLStreamHandler> clazz =
+          (Class<URLStreamHandler>) ReflectionUtils.getClassByName(customUrlStreamHandlerClassName);
+      URLStreamHandler customHandler = ReflectionUtils.newInstance(clazz);
+      url = new URL(null, sb.toString(), customHandler);
+    } else {
+      url = new URL(sb.toString());
+    }
+    return url;
+  }
+
+  private static HttpContent buildContent(
+      ApiResource.RequestMethod method, Map<String, Object> params) throws IOException {
+    if (method != ApiResource.RequestMethod.POST) {
+      return null;
```

Would it be better to throw an exception than silently omitting the content?

---

I don't think we'd want that -- it's expected that non-POST requests have no bodies. If we threw an exception here we'd have to conditionally call `buildContent()`. I don't see much of an upside right now (but I have more refactors incoming so I'll keep it in mind).

---

Makes sense. I have some paranoia about the prospect of contentful non-posts due to past experience (not at Stripe) ;)

---

Fun fact (for some definitions of "fun"): Stripe's API will happily accept parameters in both query strings and request bodies, regardless of the HTTP method being used 🙈